libp2p/Host.cpp

```diff
@@ -145,70 +141,76 @@ void Host::stop()
-            for (auto i: m_sessions)
-                if (auto p = i.second.lock())
-                    if (p->isConnected())
+        // reset ioservice (cancels all timers and allows manually polling network, below)
+        m_ioService.reset();
+
+        DEV_GUARDED(x_timers)
+            m_timers.clear();
+        
+        // shutdown acceptor
+        m_tcp4Acceptor.cancel();
+        if (m_tcp4Acceptor.is_open())
+            m_tcp4Acceptor.close();
+
+        // There maybe an incoming connection which started but hasn't finished.
+        // Wait for acceptor to end itself instead of assuming it's complete.
+        // This helps ensure a peer isn't stopped at the same time it's starting
+        // and that socket for pending connection is closed.
+        while (m_accepting)
+            m_ioService.poll();
+
+        // stop capabilities (eth: stops syncing or block/tx broadcast)
+        for (auto const& h: m_capabilities)
+            h.second->onStopping();
+
+        // disconnect pending handshake, before peers, as a handshake may create a peer
+        for (unsigned n = 0;; n = 0)
+        {
+            DEV_GUARDED(x_connecting)
+                for (auto const& i: m_connecting)
+                    if (auto h = i.lock())
                     {
-                        p->disconnect(ClientQuit);
+                        h->cancel();
                         n++;
                     }
-        if (!n)
-            break;
-
-        // poll so that peers send out disconnect packets
-        m_ioService.poll();
-    }
+            if (!n)
+                break;
+            m_ioService.poll();
+        }
+        
+        // disconnect peers
+        for (unsigned n = 0;; n = 0)
+        {
+            DEV_RECURSIVE_GUARDED(x_sessions)
+                for (auto i: m_sessions)
+                    if (auto p = i.second.lock())
+                        if (p->isConnected())
+                        {
+                            p->disconnect(ClientQuit);
+                            n++;
+                        }
+            if (!n)
+                break;
+
+            // poll so that peers send out disconnect packets
+            m_ioService.poll();
+        }
 
-    // stop network (again; helpful to call before subsequent reset())
-    m_ioService.stop();
+        // stop network (again; helpful to call before subsequent reset())
+        m_ioService.stop();
 
-    // reset network (allows reusing ioservice in future)
-    m_ioService.reset();
+        // reset network (allows reusing ioservice in future)
+        m_ioService.reset();
 
-    // finally, clear out peers (in case they're lingering)
-    RecursiveGuard l(x_sessions);
-    m_sessions.clear();
+        // finally, clear out peers (in case they're lingering)
+        RecursiveGuard l(x_sessions);
+        m_sessions.clear();
+    }
+    else
+        // allows reusing ioservice in the future
+        m_ioService.reset();
```

Comment:
This can be simplified by moving it out of `if`, as it's present in both branches